Added `torch.baddbmm` operator

riccardofelluga opened the pull request:
<details>
  <summary><b>Before submitting</b></summary>

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/pytorch-lightning/blob/main/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to update the docs?
- [x] Did you write any new necessary tests?

</details>

## What does this PR do?

This PR works towards supporting NeMo models, in particular fixes #326. I still think that the version mentioned in the issue is the inplace one, however this op is needed.

## Did you have fun?

One more op! 🚀


cc @apaz-cli @tfogal

nikitaved commented on thunder/torch/__init__.py:
```diff
@@ -3041,6 +3041,26 @@ def batch_norm(
+        utils.check_type(alpha, int)
+
+    t0 = matmul(b1, b2)
+    t1 = mul(alpha, t0)
```
no need to `mul` for the same reason.

mruberry replied:
nit: `alpha * t0`

nikitaved commented on thunder/torch/__init__.py:
```diff
@@ -3041,6 +3041,26 @@ def batch_norm(
+    a: TensorLike, b1: TensorLike, b2: TensorLike, *, beta: float = 1.0, alpha: float = 1.0, out: TensorLike = None
+) -> TensorLike:
+    utils.check(out is None, lambda: "Non-None out is not supported", NotImplementedError)
+
+    utils.check_same_dtype(a, b1, b2)
```
We will probably do shortcuts below, so we should probably check that all things are on the same device.

nikitaved commented on thunder/tests/opinfos.py:
```diff
@@ -6153,6 +6153,63 @@ def tensor_1d_sample_generator(op, device, dtype, requires_grad, **kwargs):
+    cases = (
+        ((0, 0, 0), (0, 0, 0), (0, 0, 0)),
+        ((3, 0, 5), (3, 0, 0), (3, 0, 5)),
+        ((0, 5, 6), (0, 5, 0), (0, 0, 6)),
+        ((3, 5, 6), (3, 5, 0), (3, 0, 6)),
```
Let's add cases where the matmul part actually produces something other than zero

nikitaved commented on thunder/torch/__init__.py:
```diff
@@ -3041,6 +3041,26 @@ def batch_norm(
+    utils.check(b1.ndim == 3, lambda: f"batch1 must be a 3D tensor, found {b1.ndim} instead.")
+    utils.check(b2.ndim == 3, lambda: f"batch2 must be a 3D tensor, found {b2.ndim} instead.")
+
+    if a.dtype not in dtypes.inexact_dtypes:
+        utils.check_type(beta, int)
+        utils.check_type(alpha, int)
```
Could you please elaborate on these lines?

riccardofelluga commented on thunder/torch/__init__.py:
```diff
@@ -3041,6 +3041,26 @@ def batch_norm(
+        utils.check_type(beta, int)
+        utils.check_type(alpha, int)
+
+    t0 = matmul(b1, b2)
```
That's true, however I've noticed that it broke the gradient computation. Let me see what I can do

mruberry replied:
Curious what's going on with that; also fyi @IvanYashchuk

tfogal replied:
it is odd to break gradient...

but, i would think the matmul call itself would do this optimization if possible, and it'll do so with fewer instructions. If we want to do this, I'd recommend doing it at a higher level, via (say) a constant propagation pass, which could remove the op entirely and then potentially DCE some tensors.

mruberry replied:
You're absolutely right that the logic could be handled by an optimization pass, @tfogal. I don't think we currently perform numeric simplifications. 

Also, is it true that we don't need to perform this matmul if `alpha` is zero? What if one or both of the tensors contained NaNs?

mruberry commented on thunder/torch/__init__.py:
```diff
@@ -3041,6 +3041,26 @@ def batch_norm(
 
+@torchsymbol(torch.baddbmm, is_method=True)
+def baddbmm(
+    a: TensorLike, b1: TensorLike, b2: TensorLike, *, beta: float = 1.0, alpha: float = 1.0, out: TensorLike = None
```
`, /, *,` let's not allow binding to names like `a`, `b1` and `b2`

mruberry commented on thunder/torch/__init__.py:
```diff
@@ -3041,6 +3041,26 @@ def batch_norm(
+    t0 = matmul(b1, b2)
+    t1 = mul(alpha, t0)
+
+    return add(t1, a, alpha=beta)
```
nit: `t1 + (a * beta)`

mruberry commented:
Looks really good so far, @riccardofelluga! Just some minor tweaks

IvanYashchuk commented on thunder/tests/opinfos.py:
```diff
@@ -6152,6 +6152,72 @@ def tensor_1d_sample_generator(op, device, dtype, requires_grad, **kwargs):
+        ((3, 5, 6), (3, 5, 8), (3, 8, 6)),
+    )
+
+    constants_cases = (
+        (2.0, 2.0),
+        (1.0, 0.0),
+        (0.0, 1.0),
+    )
```
`float('inf')`, `float('-inf')`, `float('nan')` are also important to check besides 0.0, 1.0, and 2.0. Could you please add them?

riccardofelluga commented:
I've added the singularities in the testing. However, I reverted back to the code without the early exits because doing so would require writing the backward implementation for those computation paths. At the moment I think we can live with it and then in the future we might want to implement some optimization pass that does this for us.  

cc @nikitaved @mruberry @tfogal

mruberry reviewed (approved):
<img src="https://media4.giphy.com/media/73Y6EEJqbyGovenlLE/giphy.gif"/>
